proxy/envoy/resources.go

```diff
@@ -230,6 +230,34 @@ type TCPRoute struct {
 	DestinationPorts  string   `json:"destination_ports,omitempty"`
 	SourceIPList      []string `json:"source_ip_list,omitempty"`
 	SourcePorts       string   `json:"source_ports,omitempty"`
+
+	// special value to retain dependent cluster definition for TCP routes.
+	clusterRef *Cluster
+}
+
+// Merge operation selects a union of two route configs prioritizing the first.
+func (rc *TCPRouteConfig) merge(that *TCPRouteConfig) *TCPRouteConfig {
```

Comment:
I think you just need to use `route.Cluster.Hostname` as keys rather than clusters. That way the the first (inbound) route takes priority.